packages/core/database/src/utils/identifiers/shortener.ts

```diff
@@ -0,0 +1,223 @@
+/**
+ * @fileoverview This file contains utility functions for shortening identifiers for use in a database schema.
+ * The functions in this file are used to generate shorter names for database tables and columns
+ * to avoid breaking the constraints of databases.
+ *
+ * IMPORTANT
+ * Any changes here that result in a different output string from any of the naming methods will
+ * cause the schema creation on Strapi bootstrap to delete data it doesn't recognize because the name
+ * is different.
+ *
+ * If there are any test failures after updating this code, it means there is a breaking change that
+ * will cause data loss, so beware; do not update the test to match your changes
+ *
+ * @internal
+ */
+
+import crypto from 'node:crypto';
+import { partition, isInteger, sumBy } from 'lodash/fp';
+
+// TODO: Names will not be shortened until this is set to a non-zero number (most likely 55)
+export const MAX_DB_IDENTIFIER_LENGTH = 0;
+
+// We can accept a number of compressible tokens up to:
+// tokens accepted = (MAX_LENGTH / (HASH_LENGTH + MIN_TOKEN_LENGTH) + (tokens * IDENTIFIER_SEPARATER.length))
+// Be aware of that when considering changing these values, we should be able to support at least 4 compressible identifiers
+
+export const HASH_LENGTH = 5;
+export const HASH_SEPARATOR = ''; // no separator is needed, we will just attach hash directly to shortened name
+export const IDENTIFIER_SEPARATOR = '_';
+export const MIN_TOKEN_LENGTH = 3;
+
+type NameToken = {
+  allocatedLength?: number;
+  name: string;
+  compressible: boolean;
+};
+
+type NameTokenWithAllocation = NameToken & { allocatedLength: number };
+
+/**
+ * Creates a hash of the given data with the specified string length as a string of hex characters
+ *
+ * @example
+ * createHash("myData", 5); // "03f85"
+ * createHash("myData", 2); // "03"
+ * createHash("myData", 1); // "0"
+ *
+ * @param data - The data to be hashed
+ * @param len - The length of the hash
+ * @returns The generated hash
+ * @throws Error if the length is not a positive integer
+ * @internal
+ */
+export function createHash(data: string, len: number): string {
+  if (!isInteger(len) || len <= 0) {
+    throw new Error(`createHash length must be a positive integer, received ${len}`);
+  }
+
+  // TODO: shake256 is based on SHA-3 and is slow, we don't care about cryptographically secure, only uniqueness and speed
+  //       investigate alternatives before releasing this. But it is only run on startup, so it should be fine.
+  const hash = crypto.createHash('shake256', { outputLength: Math.ceil(len / 2) }).update(data);
```

Comment:
we could use `xxhash` as an alternative algo, it's fast and not cryptographic
I do think we need to add package to be able to use it though

Reply:
That seems to be the best one. But the main question: is `substring(xxhash(string), 6)` still as "unique" as `shake256(string, 6)`? And even if it is, is it worth adding another package just for that?

Reply:
I mean technically this code will only be running when the user is developing something on Strapi, so the speed  shouldn't be much of a downside, it already takes some time to create and edit tables I don't think this will add much more on top